src/core/api.go

```diff
@@ -185,74 +185,24 @@ func (c *Core) SetLogger(log util.Logger) {
 //		socks://a.b.c.d:e/f.g.h.i:j
 // This adds the peer to the peer list, so that they will be called again if the
 // connection drops.
-/*
-func (c *Core) AddPeer(addr string, sintf string) error {
-	if err := c.CallPeer(addr, sintf); err != nil {
-		// TODO: We maybe want this to write the peer to the persistent
-		// configuration even if a connection attempt fails, but first we'll need to
-		// move the code to check the peer URI so that we don't deliberately save a
-		// peer with a known bad URI. Loading peers from config should really do the
-		// same thing too but I don't think that happens today
+func (c *Core) AddPeer(uri string, sourceInterface string) error {
+	u, err := url.Parse(uri)
+	if err != nil {
 		return err
 	}
-	c.config.Mutex.Lock()
-	defer c.config.Mutex.Unlock()
-	if sintf == "" {
-		for _, peer := range c.config.Current.Peers {
-			if peer == addr {
-				return errors.New("peer already added")
-			}
-		}
-		c.config.Current.Peers = append(c.config.Current.Peers, addr)
-	} else {
-		if _, ok := c.config.Current.InterfacePeers[sintf]; ok {
-			for _, peer := range c.config.Current.InterfacePeers[sintf] {
-				if peer == addr {
-					return errors.New("peer already added")
-				}
-			}
-		}
-		if _, ok := c.config.Current.InterfacePeers[sintf]; !ok {
-			c.config.Current.InterfacePeers[sintf] = []string{addr}
-		} else {
-			c.config.Current.InterfacePeers[sintf] = append(c.config.Current.InterfacePeers[sintf], addr)
-		}
+	if err := c.CallPeer(u, sourceInterface); err != nil {
+		return err
 	}
+	c.config._peers[Peer{uri, sourceInterface}] = struct{}{}
```

Comment:
It isn’t safe to modify `c.config._peers` from outside of the `*Core` actor (hence the underscore at the beginning of the name) — you should surround this with something like:
```
phony.Block(c, func() {
   …
})
```

Reply:
`CallPeer` was right to be outside of the `phony.Block` and to check the error etc as before — only the `c.config._peers` bit needed protecting.

Reply:
Should the error really be checked for `CallPeer`? What if a there is a temporary failure on the first connection?

Reply:
Well, we can either ignore the fact that the peer might not be reachable in case they come back in the future, or we can make more more immediately apparent to the user that the endpoint was dead. Not sure which is better from an ergonomic perspective.

Reply:
I think its better to return the error immediately and the admin client can just try again later.